File: model_1/CS3233_2024_Midterm/AplusB_or_AminusB_1_solution.py
import sys
from functools import lru_cache

def main():
    global L_val, R_val
    L_val, R_val = map(int, sys.stdin.readline().split())
    
    @lru_cache(maxsize=None)
    def dfs(i, carry, tl, th, tb, prev):
        if i == 32:
            return 1 if carry == 0 else 0
        
        L_i = (L_val >> i) & 1
        R_i = (R_val >> i) & 1
        
        res = 0
        low_B = L_i if tl else 0
        high_B = R_i if th else 1
        
        for b in range(low_B, high_B + 1):
            for x in [0, 1]:
                if i > 0 and x == 1 and prev == 1:
                    continue
                
                total = b + x + carry
                bit = total & 1
                new_carry = total // 2
                
                if tb:
                    if bit > R_i:
                        continue
                    new_tb = 1 if (bit == R_i) else 0
                else:
                    new_tb = 0
                
                new_tl = 1 if (tl and (b == L_i)) else 0
                new_th = 1 if (th and (b == R_i)) else 0
                
                res += dfs(i + 1, new_carry, new_tl, new_th, new_tb, b)
        return res
    
    result = dfs(0, 0, 1, 1, 1, 0)
    print(result)

if __name__ == '__main__':
    main()

File: model_1/CS3233_2024_Midterm/AplusB_or_AminusB_1_solution.cpp
#include <iostream>
#include <cstring>
using namespace std;

typedef long long ll;

ll L_val, R_val;
ll dp[33][2][2][2][2][2];

ll dfs(int i, int carry, bool tl, bool th, bool ta, int prev) {
    if (i == 32) {
        return (carry == 0) ? 1 : 0;
    }
    if (dp[i][carry][tl][th][ta][prev] != -1) 
        return dp[i][carry][tl][th][ta][prev];
    
    ll res = 0;
    int L_i = (L_val >> i) & 1;
    int R_i = (R_val >> i) & 1;

    int low_B = tl ? L_i : 0;
    int high_B = th ? R_i : 1;

    for (int b = low_B; b <= high_B; b++) {
        for (int x = 0; x < 2; x++) {
            if (i > 0 && x == 1 && prev == 1) 
                continue;

            int total = b + x + carry;
            int bit_A = total & 1;
            int new_carry = total >> 1;

            if (ta) {
                if (bit_A > R_i) 
                    continue;
            }

            bool new_tl = tl && (b == L_i);
            bool new_th = th && (b == R_i);
            bool new_ta = ta && (bit_A == R_i);

            res += dfs(i+1, new_carry, new_tl, new_th, new_ta, b);
        }
    }

    return dp[i][carry][tl][th][ta][prev] = res;
}

int main() {
    cin >> L_val >> R_val;
    memset(dp, -1, sizeof(dp));
    ll ans = dfs(0, 0, 1, 1, 1, 0);
    cout << ans << endl;
    return 0;
}